src/backend/distributed/utils/foreign_key_relationship.c

```diff
@@ -98,6 +104,108 @@ static void VisitOid(HTAB *oidVisitedMap, Oid oid);
+	LOCKMODE lockMode = ShareRowExclusiveLock;
+	Relation relation = try_relation_open(relationId, lockMode);
+	if (!RelationIsValid(relation))
+	{
+		ereport(ERROR, (errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						errmsg("relation with OID %d does not exist",
+							   relationId)));
+	}
+
+	relation_close(relation, NoLock);
+
+	bool foundInFKeyGraph = false;
+	ForeignConstraintRelationshipNode *relationshipNode =
+		GetRelationshipNodeForRelationId(relationId, &foundInFKeyGraph);
+	if (!foundInFKeyGraph)
+	{
+		/*
+		 * Relation could not be found in foreign key graph, then it has no
+		 * foreign key relationships.
+		 */
+		return NIL;
+	}
+
+	List *fKeyConnectedRelationshipNodeList =
+		GetRelationshipNodesForFKeyConnectedRelations(relationshipNode);
+	List *fKeyConnectedRelationIdList =
+		GetRelationIdsFromRelationshipNodeList(fKeyConnectedRelationshipNodeList);
+	return fKeyConnectedRelationIdList;
+}
+
+
+/*
+ * GetRelationshipNodesForFKeyConnectedRelations performs breadth-first search
+ * starting from input ForeignConstraintRelationshipNode and returns a list
+ * of ForeignConstraintRelationshipNode objects for relations that are connected
+ * to given relation node via a foreign key relationhip graph.
+ */
+static List *
+GetRelationshipNodesForFKeyConnectedRelations(
+	ForeignConstraintRelationshipNode *relationshipNode)
+{
+	HTAB *oidVisitedMap = CreateOidVisitedHashSet();
+
+	VisitOid(oidVisitedMap, relationshipNode->relationId);
+	List *relationshipNodeList = list_make1(relationshipNode);
+
+	ForeignConstraintRelationshipNode *currentNode = NULL;
+	foreach_ptr_append(currentNode, relationshipNodeList)
+	{
+		List *allNeighboursList = GetAllNeighboursList(currentNode);
+		ForeignConstraintRelationshipNode *neighbourNode = NULL;
+		foreach_ptr(neighbourNode, allNeighboursList)
+		{
+			Oid neighbourRelationId = neighbourNode->relationId;
+			if (OidVisited(oidVisitedMap, neighbourRelationId))
+			{
+				continue;
+			}
+
+			VisitOid(oidVisitedMap, neighbourRelationId);
+			relationshipNodeList = lappend(relationshipNodeList, neighbourNode);
+		}
+	}
+
+	return relationshipNodeList;
+}
+
+
+/*
+ * GetAllNeighboursList returns a list of ForeignConstraintRelationshipNode
+ * objects by concatenating both (referencing & referenced) adjacency lists
+ * of given relationship node.
+ */
+static List *
+GetAllNeighboursList(ForeignConstraintRelationshipNode *relationshipNode)
+{
+	bool isReferencing = false;
+	List *referencedNeighboursList = GetNeighbourList(relationshipNode, isReferencing);
+
+	isReferencing = true;
+	List *referencingNeighboursList = GetNeighbourList(relationshipNode, isReferencing);
+
+	/*
+	 * GetNeighbourList returns list from graph as is, so first copy it as
+	 * list_concat might invalidate it.
+	 */
+	List *allNeighboursList = list_copy(referencedNeighboursList);
+	allNeighboursList = list_concat_unique_ptr(allNeighboursList,
+											   referencingNeighboursList);
```

Comment:
One thing I about: What is the lifetime of the stuff we are returning here? When does it get invalidated? Can it happen at any time? Should we be making a deep copy in the end to into the current memory context?

Reply:
Actually we have three different places to invalidate foreign key graph:
* at CreateCitusLocalTable,
* at CreateDistributedTable,
* and after standard_ProcessUtility executes

So I couldn't think of any reasons for unexpected invalidations
But to be on the safe side, we can do a deep copy here.

Reply:
Well, InvalidateForeignKeyGraph sends an invalidation message to all backends, so it could happen at any time due to concurrent processes.

Deep copy sounds good.

Reply:
Actually, these functions are all static, so as long as we avoid AcceptInvalidationCalls it should be ok.